### ISSUE TITLE
ssl-redirect doesn't work when terminating TLS upstream

### ISSUE THREAD

**Issue by jhohertz** [jcmoraisjr/haproxy-ingress#401]
It seems that unless I am terminating TLS with haproxy-ingress itself, I do not have the option to use the ssl-redirect global OR the equivalent per-ingress annotations?

This seems to defeat the purpose of the https-to-http-port function, as there is no way, to force the redirect to https.

These comments seem to confirm this is the case: https://github.com/jcmoraisjr/haproxy-ingress/blob/master/pkg/controller/config.go#L472-L473

Can it be allowed to use the ssl-redirect feature, regardless if the internal SSL support is used?

**Comment by jhohertz** [comment]
I should note I have tested w/ 0.7.2 and 0.8-beta.3

**Comment by jcmoraisjr** [comment]
Hi, this is definitely wrong, thanks for pointing out. The template topology on v0.7 makes this a bit hard to fix but shouldn't be a problem on v0.8. Please share your relevant controller configurations regarding TLS and also what your fronting lb is doing regarding http and https access.

**Comment by jhohertz** [comment]
Sorry for the rambling response, discovered some things as I went. 0.8 should be okay I think, issue limited to 0.7.

Using [kube-ingress-aws-controller](https://github.com/zalando-incubator/kube-ingress-aws-controller) to manage AWS ALB's that forward into the haproxy-ingress, and terminate SSL. It listens on both port 80 and 443, and forwards both.

Pretty much the only relevant setting I have in the configmap is `https-to-http-port: "80"`. I've tried it "shared" with the main http port, and giving it it's own with similar effect. I am not setting up any kind of certs with the haproxy configuration.

With the 0.7 it seems like the rules are just tied to SSL terminating at the ingress vs. externally.
With 0.8 I am less clear what is happening... the ingress is being added to `/etc/haproxy/maps/_global_https_redir.map` 

So this bit from _front_http *should* be punting it over to https I think?
```
    http-request set-var(req.base) base,regsub(:[0-9]+/,/)
    http-request redirect scheme https if { var(req.base),map_beg(/etc/haproxy/maps/_global_https_redir.map,_nomatch) yes }
```
So when I tried out 0.8, I had failed to try again, setting https-to-http-port to 80. It was on it's own dedicated port. 

Using a shared port, it appears this works as intended. The problem with running on a dedicated port appears to be that the controller binds that port into the https front end, so the above rule never gets a chance to fire. (As the front end ALB controller forwards everything back to the same port... if it instead allowed us to send http and https to different ports, this would work fine.).

So in short... sorry for the distraction here... not sure what if anything could be done to patch up 0.7 to work better, but given that the 0.8 is working fairly well, I will wait for that to hit stable I think. Feel free to close this at your discretion.

**Comment by jcmoraisjr** [comment]
Hi, the current implementation wasn't prepared to run sharing the `:80` in the proxy. I'll have a look on this scenario. What's the difference between a request from `:80` and from `:443`, something like `X-Forwarded-Proto` header added with `http` / `https`? Is this approach always used and I can analyse the request based on this value? Is there something else?

**Comment by jhohertz** [comment]
Interesting, it seemed to work at first, but had to back it out back to 0.7.2 due to some issues with websockets that I haven't been able to analyse yet.

Re: the ALB in front, yes, they use the "standard" headers for this like `X-Forwarded-Proto`, doc here: https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/x-forwarded-headers.html

I'm undecided on if it's best to try to adapt this to allow a shared port for the incoming traffic, or adapt the ALB controller to allow splitting the listeners to different destination ports.

**Comment by jcmoraisjr** [comment]
Hi, adapting the ALB to use distinct ports would make things a lot easier, the topology was built with plain http and secure (offloading local or in the fronting proxy) in distinct frontends, which need distinct ports. This need to be fixed in either one or another way:

* Please confirm if distinct ports can be used here, if so I'll remove the option to share the http port and document this properly; or
* I'll have a look on this shortly if the use of distinct ports cannot be an option. Btw thanks for the detailed descriptions of the scenario you've been provided so far =)

**Comment by sschaap** [comment]
We are in the process of migrating services from colocated hardware to a GKE cloud-based setup. Using a GCP global load balancer (provisioned through a GKE ingress) gives us the ability to terminate and unwrap SSL near the user, providing a much faster handshake turnaround time (fewer asia-europe-asia bounces). However, (re)configuration of GKE ingresses is especially slow and takes up to 15 minutes per change.

I'm looking into using a GCP LB for SSL termination, then handing HTTP transactions off to a secondary internal ingress based on another technology. That would afford us both global SSL termination and faster/more flexible configuration of the internal ingress.

Now we also want our setup to redirect HTTP requests to an equivalent HTTPS URL. However, since we are behind another load balancer (effectively acting as an HTTP proxy) we can no longer use the request scheme to check whether the request is secure. Neither can we check whether the socket is SSL enabled (which I think is what `ssl_fc` does?) since the external LB unwraps the request.

Our current solution is implement redirection in each and every backend service, which is suboptimal. What that does give us, though, is an idea on how to implement this at the load balancer stage. Our redirection implementation uses `X-Forwarded-Proto` to determine the request scheme if it is defined. Otherwise, we fall back to the original request scheme. If we find that `X-Fwd-Proto` equals `http` then we redirect to `https://$host$request_uri` instead.

Just my 2 cents. Hope that helps the discussion on this issue.

**Comment by jcmoraisjr** [comment]
Hi, thank you for the detailed description! This confirms to me that we are going to the right route using the `X-Forwarded-Proto`. This is exactly the propose of the badly named `https-to-http-port` config key, which is broken on v0.8 due to a lack of information.

My proposal is to reimplement the functionality in this way:

* The admin should configure a distinct port number, say `8000`. If the port number conflict with the http one, which is `80` by default, this port number will assume the new behaviour described abobe and won't respect eg ssl-redirect config anymore;
* If `X-Forwarded-Proto` is `http`, a redirect scheme will be used to send users to https;
* If `X-Forwarded-Proto` is `https`, the request will be handled as if the ssl-offload was done in haproxy, eg adding hsts if configured and so on;
* If `X-Forwarded-Proto` doesn't match `http/https` or is missing, a 403 response is sent;
* A whitelist will be added if you want to limit which CIDRs can connect to that port.

Any other spec? Does this sounds good enough to the next and last v0.8-beta?

Thanks!

**Comment by jhohertz** [comment]
This sounds great @jcmoraisjr 

I also just wanted to mention, that I personally worked around this via implementing the redirection at the ALB in front of our haproxy ingress (see: zalando-incubator/kube-ingress-aws-controller#276), so the issue isn't pressing for me any longer. (Though I am sure it will be valuable for others!)

**Comment by sschaap** [comment]
Ah, I was already wondering what `https-to-http-port` was meant to configure :p

Your proposed approach sounds reasonable. A few questions:

- Why have a separate listen port to determine ssl-redirect behavior? Wouldn't it be sufficient to allow configuration of a "secure proxy header" annotation to tell HAproxy that `X-Forwarded-For: https` (or any other combination of header name/value) indicates a secure request? Whitelist sounds smart though.

- Do you think `HTTP 403` is the most fitting status code for this condition? If not using a separate listening port, absence of an `XFF` header could simply drop request handling back to the "normal" path? Alternatively, perhaps `HTTP 400` with a descriptive message would be more applicable.

- Need to consider the effect of redirects on health checks. Probes defined in k8s manifests under `readinessProbes` and `livenessProbes` *must* return `HTTP 200`. These requests generally do not include an `XFF` header.

Thoughts?

**Comment by sschaap** [comment]
> worked around this via implementing the redirection at the ALB in front of our haproxy ingress (see: [zalando-incubator/kube-ingress-aws-controller#276](https://github.com/zalando-incubator/kube-ingress-aws-controller/pull/276))

It looks like you essentially run a redirector backend on port 80 and relay 443 requests to your actual backends, right? We tried doing something similar with GCP LBs by provisioning one GKE ingress without TLS config, then another on the same static IP with `k8s.io/ingress.allow-http = "false"`. The HTTP-only ingress forwards to a dedicated nginx redirector instance, while the HTTPS-only ingress forwards to the respective backends.

Sadly, the HTTPS ingress did not work correctly with allow-http disabled. Additionally, while it does not create forwarding rules on port 80 of that address, the IP suddenly does start responding on port 80 with `HTTP 404` responses. I was unable to get these ingresses to work together on a single IP address.

Just putting that out there for reference.

**Comment by jhohertz** [comment]
Our setup terminates TLS at the AWS ALB, and relays to http at the haproxy ingress. The problem was it was also forwarding port 80 inward as well, and as I opened this issue, was struggling to coax haproxy to provide a proper redirect in that configuration. The change I linked causes the exterior ALB to emit the redirect out there, and haproxy now never sees it.

**Comment by jcmoraisjr** [comment]
Hi there, thank you all about sharing your thoughts and experience, this helps haproxy-ingress to be a better product! Unfortunately this is a feature I cannot use and properly test myself, we have only on-premisse k8s clusters and our workflow uses ingress and secrets to configure ssl-offload.

> Ah, I was already wondering what `https-to-http-port` was meant to configure :p

There are some badly named features in place which I want to deprecate in a backward compatible way. Feel free to share your thoughts about names which reflect this functionality in a better way. Planning to place in [this group](https://haproxy-ingress.github.io/docs/configuration/keys/#bind-port) of another binding ports.

> Your proposed approach sounds reasonable. A few questions:
> 
> * Why have a separate listen port to determine ssl-redirect behavior? Wouldn't it be sufficient to allow configuration of a "secure proxy header" annotation to tell HAproxy that `X-Forwarded-For: https` (or any other combination of header name/value) indicates a secure request? Whitelist sounds smart though.

The main problem here is that the configuration was designed in a way that :80/http and :443/https uses distinct haproxy frontends. The main reasons here are simplicity in the template and performance in the final config file due to less `if`s. The new configured port uses the :443/https frontend due to its nature.

This is however a feature that would benefit from using the same frontend for all the incoming requests. I'll have a closer look into this, all the config generator is a lot more flexible compared with past releases or even the firsts v0.8 snapshot.

> * Do you think `HTTP 403` is the most fitting status code for this condition? If not using a separate listening port, absence of an `XFF` header could simply drop request handling back to the "normal" path? Alternatively, perhaps `HTTP 400` with a descriptive message would be more applicable.

This makes sense. The normal path is in fact the right choice when both http and https-to-http shares the same port but this can only be done in a clean way if http and https shares the same frontend. Otherwise I'd use the same behaviour if the requests goes to :8000/https-to-http on a distinct port without `X-Forwarded-Proto`.

Regarding the missing header and distinct path a 400 and descriptive message sounds good.

> * Need to consider the effect of redirects on health checks. Probes defined in k8s manifests under `readinessProbes` and `livenessProbes` _must_ return `HTTP 200`. These requests generally do not include an `XFF` header.

Probes are done via kubelet which doesn't uses the ingress path, but the pod IP instead. Am I missing another external health check which cannot be covered by [healths-port](https://haproxy-ingress.github.io/docs/configuration/keys/#bind-port)?

**Comment by jcmoraisjr** [comment]
HI, #434 should fixes this issues, perhaps you can provide some feedback based on the update to the doc. The time to change is now! Planning to tag beta.5 in a few days so you can also run in your own environment. Thanks for all the help!

**Comment by sschaap** [comment]
That is some fast feedback on this issue! Several remarks:

- If `fronting-proxy-port` is configured to match `http-port` and the `XFF` header is absent, what happens exactly? I would expect that the request is handled as if it were a regular request to the HTTP frontend and that any `ssl-redirect` configuration still applies?

- It looks as though the `forwardfor` annotation configures haproxy to emit an `X-Forwarded-For` header towards its backends. However, judging from the config template (though I'm no haproxy wizard), the header is copied to X-Original-FF, then blanked and set to the IP of the haproxy instance. My guess is that the `del-header` instruction is present because the default behavior of `option forwardfor` is to append the haproxy address. I would like to note the latter that is exactly the purpose of the `XFF` header: each proxy between client and server appends their address (Ref: [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For)). This allows the application to whitelist proxy IPs while retaining access to the user IP address. Propose to change this behavior to retain client & upstream proxy addresses.

- Re: health checks I'm not entirely sure now. I think you're right in that `healthz-port` suffices for pod health checks. However, the GCE L7 load balancers perform their own service health checks on top of that if I remember correctly. If I read the [docs on this](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress#health_checks) correctly, then this health check is required to use the same port as the service itself. However, to be sure I would need to test with an actual deployment.
--
Random thoughts: if I recall correctly, then GCE LBs probe the backing service with HTTP requests at `/` which must return an `HTTP 200` response. Specifically, `HTTP 3xx` responses cause the backend to be considered "unhealthy" by the LB. The request path may be changed by defining a readiness probe on the backing container matching the container port of the service. My initial thought would be that this would require a default backend that serves `HTTP 200` responses and is exempt from `ssl-redirect` behavior. Alternatively, a magic header or well-known path could be defined to serve as a probe endpoint here. Once beta.5 is released I can test on our staging cluster.

**Comment by jcmoraisjr** [comment]
> * If `fronting-proxy-port` is configured to match `http-port` and the `XFF` header is absent, what happens exactly? I would expect that the request is handled as if it were a regular request to the HTTP frontend and that any `ssl-redirect` configuration still applies?

Do you mean `X-Forwarded-Proto` header missing instead? The request would follow the "http way of doing things" and would redirect only if ssl-redirect enforces that. `X-Forwarded-Proto` header would be defined as `http` in this case.

> * It looks as though the `forwardfor` annotation configures haproxy to emit an `X-Forwarded-For` header towards its backends. However, judging from the config template (though I'm no haproxy wizard), the header is copied to `X-Original-FF`, then blanked and set to the IP of the remote peer. My guess is that the `del-header` instruction is present because the default behavior of `option forwardfor` is to _append_ the remote address. I would like to note the latter that is exactly the purpose of the `XFF` header: each proxy between client and server appends their address (Ref: [MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Forwarded-For)). This allows the application to whitelist proxy IPs while retaining access to the user IP address. Propose to change this behavior to retain client & upstream proxy addresses.

The current behaviour was made with no L4-7 fronting proxy in mind, where the requests came from untrusted senders, because of that any `XFF` header is removed - a backup is made to `X-Original-Forwarded-For`. I'll implement another forward-for option which concatenate all of them and also change the default behaviour when the request came from a fronting-proxy.

> * Re: health checks I'm not entirely sure now. I think you're right in that `healthz-port` suffices for pod health checks. However, the GCE L7 load balancers perform their own service health checks on top of that if I remember correctly. If I read the [docs on this](https://cloud.google.com/kubernetes-engine/docs/concepts/ingress#health_checks) correctly, then this health check is required to use the same port as the service itself. However, to be sure I would need to test with an actual deployment.
> --
>   Random thoughts: if I recall correctly, then GCE LBs probe the backing service with HTTP requests at `/` which must return an `HTTP 200` response. Specifically, `HTTP 3xx` responses cause the backend to be considered "unhealthy" by the LB. The request path may be changed by defining a readiness probe on the backing container matching the container port of the service. My initial thought would be that this would require a default backend that serves `HTTP 200` responses and is exempt from `ssl-redirect` behavior. Alternatively, a magic header or well-known path could be defined to serve as a probe endpoint here.

Please provide how I'd distinguish a health check from an ordinary request, so I can provide distinct behaviour. Even better if you could configure another port/path for the LB health check, `:10253/healthz` is configured by default.

> Once beta.5 is released I can test on our staging cluster.

Great, thanks! Planning to tag in the end of the week - just after implement this and some other issues found on beta.4.

**Comment by jcmoraisjr** [comment]
So finally beta.5 is out and should be reflecting the [doc](https://haproxy-ingress.github.io/docs/configuration/keys/#fronting-proxy-port). Please let me know about your tests.

**Comment by sschaap** [comment]
Awesome. Haven't been able to test last week, but should have time in the coming few days. Will report back with results!

**Comment by cu12** [comment]
Hmm, it looks like, that without `ssl-redirect: "false"` annotation is set simple http sites still get an https  redirect. 

Sample ingress object (without any annotations)
```
spec:
  rules:
  - host: www.redacted
    http:
      paths:
      - backend:
          serviceName: svc-318c5
          servicePort: 80
        path: /
  - host: redacted
    http:
      paths:
      - backend:
          serviceName: svc-318c5
          servicePort: 80
        path: /
``` 

and

```
/etc/haproxy/maps # cat _global_https_redir.map | grep "redacted"
redacted/ yes
www.redacted/ yes
```

using beta.5 here

**Comment by jcmoraisjr** [comment]
Note that [ssl-redirect](https://haproxy-ingress.github.io/docs/configuration/keys/#ssl-redirect) has `true` as its default value. You can configure in the configmap to change globally.

ps this thread is related with issues when a fronting proxy does the ssl-offload, please file a new one if this doesn't match with your deployment and you still have any problem.

**Comment by sschaap** [comment]
Just a heads-up that we have been testing v0.8-beta.5 with its new fronting-proxy functionality. Testing continues today, will report back on results. So far, it looks like the fronting-proxy functionality works as intended. Encountered a few other issues, more details on those later today.

**Comment by sschaap** [comment]
As noted, fronting-proxy functionality works like a charm. Found some other issues however:

- If no ingress exists matching the configured ingress class (i.e. no backing services are defined), then haproxy.cfg does not contain any frontend. My expectation instead would be that all requests would be served by the default backend.

- If the last matching ingress is removed, then its backend is removed from haproxy.cfg, but the entry in `_global_http_front.map` remains. It seems like that should also be removed?

Would you prefer separate issues be filed for these items?

Other than that, our setup using GCLB -> haproxy -> deployments seems to work as it should.

---

**X-Forwarded-For**

Something I feel I should note for posterity: the latest haproxy-ingress v0.9 snapshots introduce the `forwardfor: update` setting which configures haproxy with `option forwardfor`. This is meant to make haproxy extend the `X-Forwarded-For` header with the remote peer address. HAproxy implements this not by modifying the header value, but instead by emitting a subsequent header line with its value set to the peer address. While this is strictly standards-compliant for value-list headers, many downstream server implementations fail to parse this correctly and require a single `XFF` header.

Our solution is to configure `forwardfor: ifmissing` and
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: haproxy-ingress
data:
  forwardfor: "ifmissing"

  config-backend: |
    ##
    # We want HAproxy to extend X-Fowarded-For with the remote peer address.
    # However, "option forwardfor" adds an extra XFF header line instead. While
    # this is strictly RFC-compliant, many backend servers interpret this incorrectly.
    # Therefore, we use "option forwardfor if-none" above to insert the header if it is
    # not present. Aditionally, we use the below two lines to extend XFF if already present.
    #
    # Ref: https://github.com/haproxy/haproxy/issues/44
    #
    acl xff-exists req.hdr(X-Forwarded-For) -m found
    http-request replace-header X-Forwarded-For (.*) '\1, %[src]' if xff-exists
```

**Comment by jcmoraisjr** [comment]
Thank you guys for helping me on this issue! So I can consider the original issue solved? I'll close this but feel free to update with any new info on fronting proxy config.

Regarding the new issues, yes please file new ones so this will have more visibility, can be individually tracked and will have their own lifecycle. Regarding XFF this can be made using a new value to the current config key.

Thank you again for providing such useful feedback!
